docs/proposals/pod-resource-management.md

```diff
@@ -0,0 +1,394 @@
+  +- Pod2
+  |   +- Container3
+  |   ...
+  +- ...
+  |
+  +- Bu
+  |   |
+  |   +- Pod3
+  |   |   |
+  |   |   +- Container4
+  |   |   ...
+  |   +- Pod4
+  |   |   +- Container5
+  |   |   ...
+  |   +- ...
+  |
+  +- BE
+  |   |
+  |   +- Pod5
+  |   |   |
+  |   |   +- Container6
+  |   |   +- Container7
+  |   |   ...
+  |   +- ...
+  |
+  +- System-reserved
+  |   |
+  |   +- system
+  |   +- ...
+  |
+  +- Kube-reserved 
+  |   |
+  |   +- kubelet
+  |   +- docker (optional)
+  |   +- ...
+  |
+```
+
+QoS Policy Design Decisions
+
+- This hierarchy highly prioritizes resource guarantees to the Guaranteed over Burstable and Best Effort Pods.
+- By not having a separate cgroup for the Guaranteed class, the hierarchy allows the Guaranteed pods to burst and utilize all of Node's Allocatable capacity.
+- The Best Effort and Burstable pods are strictly restricted from bursting and hogging resources and thus `G` Pods are guaranteed resource isolation.
+- Best Effort pods are treated as lowest priority. So for the `BE` QoS cgroup we set cpu shares to the lowest possible value ie.2. This ensures that the best effort containers get a relatively small share of cpu time.
+- Also we don't set any quota on the cpu resources as the containers on the `BE` pods can use any amount of free resources on the node.
+- Having memory limit of `BE` cgroup as (Allocatable - summation of memory limits of Guaranteed and Burstable pods) would result in `BE` pods becoming more susceptible to being OOM killed. As more `G` and `Bu` pods are scheduled kubelet will more likely kill `BE` pods, even if the `G` and `Bu` pods are using less than their request since we will be dynamically reducing the size of `BE` m.limit_in_bytes. But this allows for better memory guarantees to the `G` and `Bu` pods.
+
+## Implementation Plan
+
+The implementation plan is outlined in the next sections.
+We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default at least in v1.4.
+
+#### Top level Cgroups for QoS tiers
+
+Two top level cgroups for Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. All guaranteed pods cgroups are by default nested under the `Root`. So we dont create a top level cgroup for the Guaranteed class. For raw cgroup systems we would use libcontainers cgroups manager for general cgroup management(cgroup creation/destruction). But for systemd we don't have equivalent support for slice management in libcontainer yet. So we will be adding support for the same in the Kubelet. These cgroups are only created once on Kubelet initialization as a part of node setup.
+
+#### Pod level Cgroup creation and deletion (Docker runtime)
+
+- When a new pod is brought up, its QoS class is firstly determined.
+- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
```

Comment:
This will happen before kubelet performs any activity for a pod. Ideally,
we can charge all pod related activities including image pulling, volumes
creation, network setup, etc., to the pod level cgroups.

On Fri, Jun 24, 2016 at 10:45 AM, Derek Carr <notifications@github.com>
wrote:

> In docs/proposals/pod-resource-management.md
> <https://github.com/kubernetes/kubernetes/pull/26751#discussion_r68434206>
> :
>
> > +- Also we don't set any quota on the cpu resources as the containers on the `BE` pods can use any amount of free resources on the node.
> > +- Having memory limit of `BE` cgroup as (Allocatable - summation of memory limits of Guaranteed and Burstable pods) would result in `BE` pods becoming more susceptible to being OOM killed. As more `G` and `Bu` pods are scheduled kubelet will more likely kill `BE` pods, even if the `G` and `Bu` pods are using less than their request since we will be dynamically reducing the size of `BE` m.limit_in_bytes. But this allows for better memory guarantees to the `G` and `Bu` pods.
> > +
> > +## Implementation Plan
> > +
> > +The implementation plan is outlined in the next sections.
> > +We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default at least in v1.4.
> > +
> > +#### Top level Cgroups for QoS tiers
> > +
> > +Two top level cgroups for Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. All guaranteed pods cgroups are by default nested under the `Root`. So we dont create a top level cgroup for the Guaranteed class. For raw cgroup systems we would use libcontainers cgroups manager for general cgroup management(cgroup creation/destruction). But for systemd we don't have equivalent support for slice management in libcontainer yet. So we will be adding support for the same in the Kubelet. These cgroups are only created once on Kubelet initialization as a part of node setup.
> > +
> > +#### Pod level Cgroup creation and deletion (Docker runtime)
> > +
> > +- When a new pod is brought up, its QoS class is firstly determined.
> > +- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
>
> will this happen before or after image pulling?
>
> /cc @vishh <https://github.com/vishh> - I would like to be able to in the
> future charge image pulling to the pod cgroup.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26751/files/86f33723eebe28602e420e4118a1a13e986b6c79#r68434206>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvIKCpbqYXhEBU8FuM31nc8RF73t0Lmks5qPBfKgaJpZM4ItEyP>
> .
>